### PR TITLE
fix: No user lookup on dotcom by email

### PR DESCRIPTION
**Pull request by dadlerj**
See the conversation at
https://sourcegraph.slack.com/archives/C89KCDK5J/p1550630417010700?thread_ts=1550628877.010400&cid=C89KCDK5J

### CODE REVIEW

File: cmd/frontend/graphqlbackend/user.go [hunk 1]
```diff
@@ -22,6 +23,12 @@ func (r *schemaResolver) User(ctx context.Context, args struct {
 	Email    *string
 }) (*UserResolver, error) {
 	if args.Email != nil {
+		// 🚨 SECURITY: Only admins are allowed to access the email address on Sourcegraph.com.
+		if envvar.SourcegraphDotComMode() {
```
Comment (tsenart) [inline cmd/frontend/graphqlbackend/user.go]: Would it hurt that we prevent non admins looking up a user by username as well in dot-com? It'd be just moving this check outside the if block.

Comment (KattMingMing) [reply cmd/frontend/graphqlbackend/user.go]: For on-prem instances I don't see an issue for users to query by username or email address.

Hypothetical Sourcegraph extension: An extension author wants to pull a list of all teammates on their Sourcegraph on-prem instance via email and then use it to identify committers on their team.

Comment (tsenart) [reply cmd/frontend/graphqlbackend/user.go]: I wasn't talking about on-prem.

Comment (sqs) [reply cmd/frontend/graphqlbackend/user.go]: It turned out that restricting lookup by username on Sourcegraph.com caused #2480, which prevented users from viewing their settings, profile, or subscriptions. Fixed in https://github.com/sourcegraph/sourcegraph/pull/2522 and added tests.

I am not aware of any harmful security implications of allowing this again, given that every other similar service I can think of allows it.

Comment (tsenart) [reply cmd/frontend/graphqlbackend/user.go]: Ouch. Thanks for the fix @sqs (and for adding tests!)

Comment (tsenart) [reply cmd/frontend/graphqlbackend/user.go]: >  I am not aware of any harmful security implications of allowing this again, given that every other similar service I can think of allows it.

It's true that many services allow this, but from first principles, it's not very hard to enumerate a list of users via guessable usernames and/or usernames that are reused across several services.

How would you compare this risk to enumerating / identifying users by e-mail address?

**Pull request merged**